plugins/inputs/intel_powerstat/intel_powerstat.go

```diff
@@ -1,915 +1,1218 @@
-		}
-		return
+	p.needsTimeRelatedMsr = needsTimeRelatedMsr(p.CPUMetrics)
+
+	p.needsRapl = needsRapl(p.PackageMetrics)
+	p.needsMsrPackage = needsMsr(p.PackageMetrics)
+
+	p.cpuMetricsMap = map[string]func(telegraf.Accumulator, int, int, int){
+		cpuC0StateResidency.String():        p.addCPUC0StateResidency,
+		cpuC1StateResidency.String():        p.addCPUC1StateResidency,
+		cpuC3StateResidency.String():        p.addCPUC3StateResidency,
+		cpuC6StateResidency.String():        p.addCPUC6StateResidency,
+		cpuC7StateResidency.String():        p.addCPUC7StateResidency,
+		cpuBusyFrequency.String():           p.addCPUBusyFrequency,
+		cpuBusyCycles.String():              p.addCPUBusyCycles,
+		cpuC0SubstateC01Percent.String():    p.addCPUC0SubstateC01Percent,
+		cpuC0SubstateC02Percent.String():    p.addCPUC0SubstateC02Percent,
+		cpuC0SubstateC0WaitPercent.String(): p.addCPUC0SubstateC0WaitPercent,
+	}
+
+	p.packageMetricMap = map[string]func(telegraf.Accumulator, int){
+		packageCurrentPowerConsumption.String():     p.addCurrentPackagePower,
+		packageCurrentDramPowerConsumption.String(): p.addCurrentDramPower,
+		packageThermalDesignPower.String():          p.addThermalDesignPower,
+		packageCPUBaseFrequency.String():            p.addCPUBaseFrequency,
+		packageTurboLimit.String():                  p.addMaxTurboFreqLimits,
 	}
-	p.logOnce["intel_uncore_frequency"] = nil
-	p.readUncoreFreq("initial", socketID, die, acc)
-	p.readUncoreFreq("current", socketID, die, acc)
+
+	return nil
 }
 
-func (p *PowerStat) readUncoreFreq(typeFreq string, socketID string, die string, acc telegraf.Accumulator) {
-	fields := map[string]interface{}{}
-	if typeFreq == "current" {
-		if p.areCoreMetricsEnabled() && p.msr.isMsrLoaded() {
-			p.logOnce[socketID+"msr"] = nil
-			cpuID, err := p.GetCPUIDFromSocketID(socketID)
-			if err != nil {
-				p.Log.Debugf("Error while reading socket ID: %v", err)
-				return
-			}
-			actualUncoreFreq, err := p.msr.readSingleMsr(cpuID, msrUncorePerfStatusString)
-			if err != nil {
-				p.Log.Debugf("Error while reading %s: %v", msrUncorePerfStatusString, err)
-				return
-			}
-			actualUncoreFreq = (actualUncoreFreq & 0x3F) * 100
-			fields["uncore_frequency_mhz_cur"] = actualUncoreFreq
-		} else {
-			err := errors.New("Error while trying to read MSR (probably msr module was not loaded), uncore_frequency_mhz_cur metric will not be collected")
-			if val := p.logOnce[socketID+"msr"]; val == nil || val.Error() != err.Error() {
-				p.Log.Errorf("%v", err)
-				// Remember that specific error occurs for socketID to omit logging next time
-				p.logOnce[socketID+"msr"] = err
-			}
+// parsePackageMetrics takes a slice of package metrics. It ensures all metrics in the slice
+// are supported package-specific, and there are no duplicates. If metrics slice is nil, then
+// default package metrics are provided.
+func parsePackageMetrics(metrics []string) ([]string, error) {
+	if metrics == nil {
```

Comment:
We usually do `if len(metrics) == 0` for lists... But not a must...

Reply:
I generally agree on this check as the proper way to check if an slice has been initialized. And in this particular case it has a heavier reason to check against `nil`. The config option `package_metrics` behaves as follows:

- `package_metrics = []` means disable package metrics. In our code `PackageMetrics != nil` but `len(PackageMetrics) == 0`.
- No package metrics in config means gather default package metrics: `"current_power_consumption"`, `"current_dram_power_consumption"`, and `"thermal_design_power"`.

This was to keep backward compatibility with the previous version of this plugin. If you don't mind I would like to keep it this way. However I will make sure to add comments to make this more clear :)